### ISSUE TITLE
ngsw.json navigationUrls breaks all service worker versions < 6.0.0

### ISSUE THREAD

**Issue by Serginho** [angular/angular#27209]
# 🐞 bug report

### Affected Package
@angular/service-worker

### Is this a regression?
5.x Works
This bug was introduced in version 6

### Description
Updating from angular 5.2.x to 7.x I run across with this error (See exception section)

The problem is the new field `navidationUrls` in the `ngsw.json` introduced in version 6.

```
"navigationUrls": [
    {
      "positive": true,
      "regex": "^\\/.*$"
    },
    {
      "positive": false,
      "regex": "^\\/(?:.+\\/)?[^/]*\\.[^/]*$"
    },
    {
      "positive": false,
      "regex": "^\\/(?:.+\\/)?[^/]*__[^/]*$"
    },
    {
      "positive": false,
      "regex": "^\\/(?:.+\\/)?[^/]*__[^/]*\\/.*$"
    }
  ]
```

When the new service worker call `initialize()` method, loads the old manifests from DatabaseCache, iterates over them,  creates de AppVersion object, as you can see here:

https://github.com/angular/angular/blob/926367150afa9138eb6f469dd111112339a07784/packages/service-worker/worker/src/driver.ts#L494-L503

In the constructor of AppVersion, old manifests (generated in the version 5.x) don't have `navigationUrls` field and breaks here:

https://github.com/angular/angular/blob/926367150afa9138eb6f469dd111112339a07784/packages/service-worker/worker/src/app-version.ts#L89-L94

The service worker is not checking the possibility of having version without that field. 

Maybe... it's assuming caches are deleted here. But caches are only deleted if manifest can't be fetched or si configVersion is not 1, but angular-cli 7 generate configVersion always with 1

https://github.com/angular/angular/blob/926367150afa9138eb6f469dd111112339a07784/packages/service-worker/worker/src/driver.ts#L754-L759

## 🔬 Minimal Reproduction
Not needed. Issue is explained.

## 🔥 Exception or Error
```
Driver state: SAFE_MODE (Initialization failed due to error: Cannot read property 'filter' of undefined
TypeError: Cannot read property 'filter' of undefined
    at new AppVersion (https://web.pre.mx.tulotero.net/ngsw-worker.js:1375:57)
    at https://web.pre.mx.tulotero.net/ngsw-worker.js:2292:49
    at Array.forEach (<anonymous>)
    at Driver.<anonymous> (https://web.pre.mx.tulotero.net/ngsw-worker.js:2287:40)
    at Generator.next (<anonymous>)
    at fulfilled (https://web.pre.mx.tulotero.net/ngsw-worker.js:1775:62))
```

## 🌍  Your Environment

**Angular Version:** 
7.0.3
Migrating 5.2.11 to 7.0.3

**Comment by gkalpak** [comment]
Ooops, good catch! It might have been safer to increment `configVersion`, but it seemed too heavy-handed.

One way to fix it would be moving the defaults from [build time][1] to the [runtime][2].
(I am not thrilled about this, because it would mean that the cached manifest would further deviate from `ngsw.json`. But I am even less thrilled with the alternative of keeping the defaults in both places 😞)

@Serginho, would you be interested in submitting a PR?

[1]: https://github.com/angular/angular/blob/a72250bacea6e29dcb5f638a9ce39f292b4310fb/packages/service-worker/config/src/generator.ts#L39
[2]: https://github.com/angular/angular/blob/926367150afa9138eb6f469dd111112339a07784/packages/service-worker/worker/src/app-version.ts#L89-L94

**Comment by Serginho** [comment]
I've been trying to move from build time to runtime, but `baseHref`doesn't exist at runtime to process default urls.
https://github.com/angular/angular/blob/a72250bacea6e29dcb5f638a9ce39f292b4310fb/packages/service-worker/config/src/generator.ts#L39

Another option is in the `setupUpdate`method do:
 - Create a setter method of `navigationUrls`.
 - Create a method `hasNavigationUrls`.
 - Iterate over the old versions checking if navigationUrls should be added.

https://github.com/angular/angular/blob/926367150afa9138eb6f469dd111112339a07784/packages/service-worker/worker/src/driver.ts#L750-L772

@gkalpak What's your opinion? Any better idea?

**Comment by Serginho** [comment]
What I wanted to say is if it's better avoid duplicating the logic, maybe we can generate default navigation urls in the manifest for backward compatibility. And in the `setupUpdate` do something like that:

```typescript
// Iterate over versions checking if navigationUrls should be added
this.versions.forEach(version => {
       if(!version.hasNavigationUrls()) {
           version.setNavigationUrls(manifest.defaultNavigationUrls);
       }
}
```

Generator:
```typescript
defaultNavigationUrls: processNavigationUrls(this.baseHref, null)
```

**Comment by gkalpak** [comment]
Oh, I see. It is definitely an option. I am not thrilled about it, since the manifest has to include `defaultNavigationUrls` which is not needed for itself but for other manifests.

<sub>(I already wish I would have increased `configVersion` and called it a breaking change :grin:)</sub>

**Comment by Serginho** [comment]
Here you are the fix.
I decided to implement the last option for seveal reasons:
- No duplicate code between `Generator` and `AppVersion' to process `navigationUrls`.
- Users are not affected in the `ngsw-config.json` definition, `defaultNavigationUrls` are inserted in the manifest at compiling time.

**Comment by gkalpak** [comment]
Thx, @Serginho 👍

As I mentioned above, I am not too happy with this approach. If nothing else, a lot of things have to change in a lot of places (as can be seen in the diff in #27244).

The more I think about it, the more I believe the right approach is to use some fallback values in the `AppVersion` constructor. I also realized two things:

1. These default values do not have to much the ones we have in [generator.ts][1], because those defaults might change in the future. For `AppVersion`s without `navigationUrls`, we need to match the behavior before 08325aaff, since that is the behavior that users would expect from manifest versions deployed back then.

2. The default URls do not rely on `baseHref`, so we don't need to worry about it in the RegExps.

So, what I am proposing is to use the following default `navigationUrls` [here][2]. (AFAICT, this should be the only change that is needed in order to remain backwards compatible.)

```ts
manifest.navigationUrls = manifest.navigationUrls || [
  {positive: true, regex: '^/.*$'},
  {positive: false, regex: '^/(?:.+/)?[^/]*\\.[^/]*$'},
  {positive: false, regex: '^/.*__'},
];
```

(Note, since we can directly manipulate the RegExps (and don't have to go through globs), we can simplify the `__` check.)

WDYT?

[1]: https://github.com/angular/angular/blob/dc300c5c4172e1f8a6f3f84da6797e427e3626c0/packages/service-worker/config/src/generator.ts#L14-L19
[2]: https://github.com/angular/angular/blob/dc300c5c4172e1f8a6f3f84da6797e427e3626c0/packages/service-worker/worker/src/app-version.ts#L89

**Comment by Serginho** [comment]
@gkalpak I updated the code to a simpler solution, but hardcoded navigationUrls.

**Comment by gkalpak** [comment]
Fixed in #27244.
